cachi2/core/package_managers/gomod.py

```diff
@@ -892,13 +891,101 @@ def go_list_deps(pattern: Literal["./...", "all"]) -> Iterator[ParsedPackage]:
+    version_resolver: "ModuleVersionResolver",
+) -> tuple[ParsedModule, list[ParsedModule]]:
+    """
+    Identify and parse the main module and all workspace modules, if they exist.
+
+    :return: A tuple containing the main module and a list of workspaces
+    """
+    modules_json_stream = go([*go_list, "-m", "-json"], run_params).rstrip()
+    main_module_dict, workspace_dict_list = _process_modules_json_stream(
+        app_dir, modules_json_stream
+    )
+
+    main_module_path = main_module_dict["Path"]
+    main_module_version = version_resolver.get_golang_version(main_module_path, app_dir)
+
+    main_module = ParsedModule(
+        path=main_module_path,
+        version=main_module_version,
+        main=True,
+    )
+
+    workspace_modules = [
+        _parse_workspace_module(app_dir, workspace_dict, main_module_version)
+        for workspace_dict in workspace_dict_list
+    ]
+
+    return main_module, workspace_modules
+
+
+def _process_modules_json_stream(
+    app_dir: RootedPath, modules_json_stream: str
+) -> tuple[ModuleDict, list[ModuleDict]]:
+    """Process the json stream returned by "go list -m -json".
+
+    The stream will contain the module currently being processed, or a list of all workspaces in
+    case a go.work file is present in the repository.
+
+    :param app_dir: the path to the module directory
+    :param modules_json_stream: the json stream returned by "go list -m -json"
+    :return: A tuple containing the main module and a list of workspaces
+    """
+    module_list = []
+    main_module = None
+
+    for module in load_json_stream(modules_json_stream):
+        if module["Dir"] == str(app_dir):
+            main_module = module
+        else:
+            module_list.append(module)
+
+    # should never happen, since the main module will always be a part of the json stream
+    if not main_module:
+        raise RuntimeError('Failed to find the main module info in the "go list -m -json" output.')
+
+    return main_module, module_list
+
+
+def _parse_workspace_module(
+    app_dir: RootedPath, module: ModuleDict, main_module_version: str
+) -> ParsedModule:
+    """Create a ParsedModule from a listed workspace.
+
+    The replacement info returned will always be relative to the module currently being processed.
+    """
+    # We can't use pathlib since corresponding method PurePath.relative_to('foo', walk_up=True)
+    # is available only in Python 3.12 and later.
+    relative_dir = os.path.relpath(module["Dir"], app_dir)
+
+    # We need to prepend "./" to a workspace that is direct child of app_dir so it can be treated
+    # the same way as a locally replaced module when converting it into a Module.
+    if not relative_dir.startswith("."):
+        relative_dir = f"./{relative_dir}"
+
+    replaced_module = ParsedModule(path=relative_dir)
+
+    return ParsedModule(
+        path=module["Path"],
+        version=main_module_version,
+        replace=replaced_module,
```

Comment:
By setting `replace` and giving the `replaced_module` a relative path, the version eventually ends up getting detected by `version_resolver.get_golang_version(workspace_module_name, workspace_module_path)`?

(It's a little confusing, but should be correct. Maybe just don't set `version=main_module_version` to reduce confusion)

Reply:
Tested locally, it does work. Would be worth adding to the integration test (can be a follow-up)

```bash
git clone https://github.com/cachito-testing/cachi2-gomod
cd cachi2-gomod

git tag workspace_modules/hello/v1.1.0
git tag workspace_modules/hi/hiii/v1.2.0
git tag workspace_modules/echo/v4.2.0

cachi2 --log-level debug fetch-deps '{"type": "gomod", "path": "workspace_modules/hello"}'
jq < cachi2-output/bom.json '.components[].purl | select(test("workspace_modules.*type=module"))' -r
```

```text
pkg:golang/github.com/cachito-testing/cachi2-gomod/workspace_modules/echo@v4.2.0?type=module
pkg:golang/github.com/cachito-testing/cachi2-gomod/workspace_modules/hello@v1.1.0?type=module
pkg:golang/github.com/cachito-testing/cachi2-gomod/workspace_modules/hi/hiii@v1.2.0?type=module
```

Reply:
I'll implement the integration test as a follow up, then.